### PR TITLE
Add script to print block numbers for BSS ordering

### PR DESCRIPTION
**Pull request by cadmic**
Recall from [IDO BSS ordering](https://hackmd.io/@Roman971/BJ2DOyhBa) that the compiler assigns "block numbers" to a bunch of things (including global variables) and that BSS is ultimately ordered by these block numbers mod 256. This script shows these block numbers to help get an idea of what can affect BSS ordering, both with and without `-g3`.

Example for debug ROM: `tools/ido_block_numbers.py src/code/z_collision_check.c | less`. This shows where BSS wraps in the debug version between `modelPos` and `worldPos`:
```
block [mod 256]: linkage  type        class      name
...
      507 [251]:          end         text       Collider_SetTrisDim
      508 [252]:          static      data       sBssDummy11
      509 [253]:          static      data       sBssDummy12
      510 [254]: extern   proc        text       Collider_UpdateSpheres
      511 [255]:          static      data       modelPos
      512 [  0]:          static      data       worldPos
      513 [  1]: extern   proc        undefined  Matrix_MultVec3f
      514 [  2]:          end         text       Collider_UpdateSpheres
      515 [  3]:          static      data       sparkInit
      516 [  4]:          end         text       CollisionCheck_SpawnRedBlood
...
```

For a -g3 example: `tools/ido_block_numbers.py -v gc-eu-mq src/boot/idle.c | less`. This shows how the BSS variables are "split" because half of them are declared in a header and the other half aren't, causing them to get reordered when we don't want them to:

```
block [mod 256]: linkage  type        class      name
...
     3853 [ 13]: extern   global      nil        gBuildTeam
     3854 [ 14]: extern   global      nil        gBuildDate
     3855 [ 15]: extern   global      nil        gBuildMakeOption
     3856 [ 16]: extern   global      bss        gPiMgrCmdQueue
     3857 [ 17]: extern   global      bss        gViConfigMode
     3858 [ 18]: extern   global      bss        gViConfigModeType
     3859 [ 19]: extern   global      nil        __osPiAccessQueue
     3860 [ 20]: extern   global      nil        __Dom1SpeedParam
     3861 [ 21]: extern   global      nil        __Dom2SpeedParam
...
     4035 [195]: extern   global      nil        gAudioHeap
     4036 [196]: extern   global      nil        gSystemHeap
     4037 [197]:          file        text       src/boot/idle.c
     4038 [198]: extern   global      bss        gMainThread
     4039 [199]: extern   global      bss        sMainStack
     4040 [200]: extern   global      bss        sMainStackInfo
     4041 [201]: extern   global      bss        sPiMgrCmdBuff
     4042 [202]: extern   proc        text       Main_ThreadEntry
     4043 [203]:          block       text
     4044 [204]: extern   proc        undefined  DmaMgr_Init
...
```

### CODE REVIEW

**Review by fig02 — APPROVED** [review]
once again have nothing to say on the implementation of the script, but think the idea of it is fine

**Review by Dragorn421 — APPROVED** [review]
Would only suggest a clearer name for the script, maybe `symtab_ido.py` ?

**Comment by fig02** [comment]
its probably cause im not as in-tune with the technical details of all of this, but a name like that tells me less about what it does

**Comment by Dragorn421** [comment]
symtab is "symbol table"
but block_numbers is fine

**Review by Dragorn421 — APPROVED** [review]

**Comment by fig02** [comment]
yeah i knew thats what it meant, but "ido symtab" doesnt tell me anything about what the script does, is what i meant.
maybe ido_symtab_parser is along the lines of what you are saying

**Comment by fig02** [comment]
@cadmic do you have intentions on changing the script name or no

**Comment by cadmic** [comment]
Hm, technically this script only prints block numbers and not any of the other information in the symbol table, although maybe it could be extended in the future. "Block numbers" also might not be correct, technically it's a "dense number" according to the docs and Roman's doc calls it a "block index" so I think I merged the two in my head, but I still like "block number" better than either of those names.

So in summary, I think the name is fine for now but I'm not attached to it

**Comment by Dragorn421** [comment]
it's really alright
merge whenever

**Comment by fig02** [comment]
think colloquially its been called block numbers for a while.
I think being more specific like symtab_block_numbers, or ido_block_numbers or something could be beneficial. cause just "block_numbers" is vague.

anyway yeah this isnt a huge deal, I just think being a little bit more descriptive could be nice :D

**Comment by cadmic** [comment]
also note that this conflicts with #1824 because of the makefile command-line parsing, but the fix is easy:

```diff
--- a/tools/block_numbers.py
+++ b/tools/block_numbers.py
@@ -232,19 +232,8 @@ def find_compiler_command_line(filename, oot_version):
         sys.exit(1)

     # Assume command line is of the form:
-    # python3 tools/asm_processor/build.py [COMPILER] -- [ASM_PROCESSOR_ARGS] -- [COMPILER_ARGS]
-    grouped = itertools.groupby(asm_processor_command_line, lambda x: x != "--")
-    result = [list(group) for equals_delim, group in grouped if equals_delim]
-    if len(result) != 3:
-        print(
-            f'Could not parse asm_processor command line: {" ".join(asm_processor_command_line)}',
-            file=sys.stderr,
-        )
-        sys.exit(1)
-
-    compiler = result[0][-1]
-    compiler_args = result[2]
-    compiler_command_line = [compiler] + compiler_args
+    # tools/reencode.sh [COMPILER] [COMPILER_ARGS]
+    compiler_command_line = asm_processor_command_line[1:]

     print(f'Command line: {" ".join(compiler_command_line)}', file=sys.stderr)
     return [compiler] + compiler_args
```
(maybe with a better name than `asm_processor_command_line`)

**Comment by cadmic** [comment]
alright, I'll change it to `ido_block_numbers.py`

**Comment by fig02** [comment]
would you prefer merging now and fixing later, or waiting for #1824. That one will probably be merged soon too if I had to guess

**Comment by cadmic** [comment]
I'm fine waiting, it's pretty easy for me to checkout this script into whatever branch I need, and I don't think other people are clamoring to use it

**Comment by fig02** [comment]
happened faster than i expected-- 1824 is merged and the fixes can be made now
